core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -99,12 +113,24 @@ public int read() throws IOException {
     if (mPosition == mLength) { // at end of file
       return -1;
     }
-    updateStream();
-    int result = mBlockInStream.read();
-    if (result != -1) {
-      mPosition++;
-    }
-    return result;
+
+    CountingRetry retry = new CountingRetry(MAX_WORKERS_TO_RETRY);
+    IOException lastException = null;
+    do {
+      updateStream();
+      try {
+        int result = mBlockInStream.read();
+        if (result != -1) {
+          mPosition++;
+        }
+        return result;
+      } catch (UnavailableException | DeadlineExceededException | ConnectException e) {
+        lastException = e;
+        handleRetryableException(mBlockInStream, e);
+        mBlockInStream = null;
+      }
+    } while (retry.attemptRetry());
+    throw lastException;
```

Comment:
Is there any way to make these retries cleaner? Just in terms of LOC added, its more than double.

Reply:
That can be hard. The original code is 6 lines. A retry logic contains at least a while loop(2 lines), a try catch block (3 lines), and initialize retry policy (1 line). That's already doubling the lines. The other things include resetting the stream, calling the exception handler, and rethrow the exception, they are all needed for good reasons and I don't think we want to trade them for brevity. The only thing I can remove on top of my mind is line 116 :).